File: tools/aot/node/naplps_demo.js
// naplps_demo.js — Load the NAPLPS NativeAOT library via koffi FFI.
//
// Usage:
//     npm install                                       # once
//     node naplps_demo.js <input.nap> <output.png> [w] [h]
//
// The node.exe finds the DLL via the path passed to koffi.load. On Windows,
// koffi walks the PATH and the given directory; we pass the absolute path so
// no env config is needed.

const fs = require('node:fs');
const path = require('node:path');
const os = require('node:os');
const koffi = require('koffi');

function libraryPath()
{
    const publishDir = path.resolve(__dirname, '..', 'publish');
    switch (os.platform())
    {
        case 'win32':  return path.join(publishDir, 'NAPLPS.dll');
        case 'darwin': return path.join(publishDir, 'libNAPLPS.dylib');
        default:       return path.join(publishDir, 'libNAPLPS.so');
    }
}

function main()
{
    const args = process.argv.slice(2);
    if (args.length < 2)
    {
        console.error(`usage: ${process.argv[1]} <input.nap> <output.png> [width] [height]`);
        process.exit(2);
    }

    const [inPath, outPath] = args;
    const width  = args[2] ? parseInt(args[2], 10) : 1024;
    const height = args[3] ? parseInt(args[3], 10) : 768;

    const libPath = libraryPath();
    if (!fs.existsSync(libPath))
    {
        console.error(`NAPLPS library not found at ${libPath}. Run tools/aot/publish.ps1 first.`);
        process.exit(1);
    }

    const lib = koffi.load(libPath);

    // FFI signatures match tools/aot/include/naplps.h.
    const napBytes    = koffi.pointer(koffi.types.uint8);
    const outBuf      = koffi.pointer(koffi.types.uint8);
    const version     = lib.func('int32_t naplps_version(uint8_t* out_buf, int32_t out_buf_len)');
    const cmdCount    = lib.func('int32_t naplps_command_count(uint8_t* bytes, int32_t len)');
    const errCount    = lib.func('int32_t naplps_error_count(uint8_t* bytes, int32_t len)');
    const renderPng   = lib.func('int32_t naplps_render_png(uint8_t* bytes, int32_t len, int32_t w, int32_t h, uint8_t* out_buf, int32_t out_buf_len)');

    const versionBuf = Buffer.alloc(32);
    const vlen = version(versionBuf, 32);
    if (vlen < 0) { console.error(`naplps_version failed: ${vlen}`); process.exit(1); }
    console.log(`NAPLPS library version: ${versionBuf.slice(0, vlen).toString('ascii')}`);

    const nap = fs.readFileSync(inPath);
    console.log(`Loaded ${inPath} (${nap.length} bytes)`);

    const nCmds = cmdCount(nap, nap.length);
    const nErrs = errCount(nap, nap.length);
    console.log(`Parsed ${nCmds} commands, ${nErrs} errors`);
    if (nCmds < 0) { console.error('parse failed'); process.exit(1); }

    // Query required size, then render.
    const required = renderPng(nap, nap.length, width, height, null, 0);
    if (required < 0) { console.error(`render failed (query): ${required}`); process.exit(1); }

    const pngBuf = Buffer.alloc(required);
    const written = renderPng(nap, nap.length, width, height, pngBuf, required);
    if (written < 0) { console.error(`render failed: ${written}`); process.exit(1); }

    fs.writeFileSync(outPath, pngBuf.slice(0, written));
    console.log(`Wrote ${outPath} (${written} bytes, ${width}x${height})`);
}

main();

File: tools/aot/ruby/naplps_demo.rb
#!/usr/bin/env ruby
# tools/aot/ruby/naplps_demo.rb
#
# Load the NAPLPS NativeAOT library via the `ffi` gem, render a .nap file to
# PNG.
#
# Install the gem once:
#     gem install ffi
#
# Usage:
#     ruby naplps_demo.rb <input.nap> <output.png> [width] [height]

require 'ffi'

module Naplps
  extend FFI::Library

  def self.library_path
    dir = File.expand_path('../publish', __dir__)
    case RUBY_PLATFORM
    when /mswin|mingw|cygwin|ucrt/ then File.join(dir, 'NAPLPS.dll')
    when /darwin/                  then File.join(dir, 'libNAPLPS.dylib')
    else                                File.join(dir, 'libNAPLPS.so')
    end
  end

  ffi_lib library_path

  # Signatures match tools/aot/include/naplps.h.
  attach_function :naplps_version,       [:pointer, :int32], :int32
  attach_function :naplps_command_count, [:pointer, :int32], :int32
  attach_function :naplps_error_count,   [:pointer, :int32], :int32
  attach_function :naplps_render_png,    [:pointer, :int32, :int32, :int32, :pointer, :int32], :int32
end

def main(argv)
  if argv.length < 2
    warn "usage: #{$PROGRAM_NAME} <input.nap> <output.png> [width] [height]"
    return 2
  end

  in_path  = argv[0]
  out_path = argv[1]
  width    = (argv[2] || 1024).to_i
  height   = (argv[3] || 768).to_i

  version_buf = FFI::MemoryPointer.new(:uint8, 32)
  vlen = Naplps.naplps_version(version_buf, 32)
  if vlen < 0
    warn "naplps_version failed: #{vlen}"
    return 1
  end
  puts "NAPLPS library version: #{version_buf.read_string(vlen)}"

  nap_bytes = File.binread(in_path)
  nap_len = nap_bytes.bytesize
  puts "Loaded #{in_path} (#{nap_len} bytes)"

  nap_ptr = FFI::MemoryPointer.new(:uint8, nap_len)
  nap_ptr.write_bytes(nap_bytes)

  n_cmds = Naplps.naplps_command_count(nap_ptr, nap_len)
  n_errs = Naplps.naplps_error_count(nap_ptr, nap_len)
  puts "Parsed #{n_cmds} commands, #{n_errs} errors"
  return 1 if n_cmds < 0

  # Query required PNG size, then render.
  required = Naplps.naplps_render_png(nap_ptr, nap_len, width, height, nil, 0)
  if required < 0
    warn "render failed (query): #{required}"
    return 1
  end

  png_ptr = FFI::MemoryPointer.new(:uint8, required)
  written = Naplps.naplps_render_png(nap_ptr, nap_len, width, height, png_ptr, required)
  if written < 0
    warn "render failed: #{written}"
    return 1
  end

  File.binwrite(out_path, png_ptr.read_bytes(written))
  puts "Wrote #{out_path} (#{written} bytes, #{width}x#{height})"
  0
end

exit(main(ARGV))
